pydantic/type_adapter.py

```diff
@@ -203,35 +208,98 @@ def __init__(
                 code='type-adapter-config-unused',
             )
 
-        config_wrapper = _config.ConfigWrapper(config)
-
-        core_schema: CoreSchema
+        self._type = type
+        self._config = config
+        self._parent_depth = _parent_depth
+        if module is None:
+            f = sys._getframe(1)
+            self._module_name = cast(str, f.f_globals.get('__name__', ''))
+        else:
+            self._module_name = module
+
+        if not self._defer_build():
+            # Immediately initialize the core schema, validator and serializer
+            self._dive_parent(1)  # +1 for this __init__
+            _, _, _ = (self.core_schema, self.validator, self.serializer)
+
+    def _dive_parent(self, depth: int) -> Self:
+        if self._parent_depth is not None:
+            self._parent_depth += depth
+        return self
+
+    @cached_property
+    def core_schema(self) -> CoreSchema:
+        """The pydantic-core schema used to build the SchemaValidator and SchemaSerializer."""
         try:
-            core_schema = _getattr_no_parents(type, '__pydantic_core_schema__')
+            return _getattr_no_parents(self._type, '__pydantic_core_schema__')
         except AttributeError:
-            core_schema = _get_schema(type, config_wrapper, parent_depth=_parent_depth + 1)
+            self._dive_parent(2)  # +2 for @cached_property and self.core_schema
+            parent_depth = self._parent_depth
+            assert parent_depth is not None
+            self._parent_depth = None  # No need to track parent depth anymore
 
-        core_config = config_wrapper.core_config(None)
-        validator: SchemaValidator
+            return _get_schema(self._type, self._config_wrapper, parent_depth=parent_depth)
+
+    @cached_property
+    def validator(self) -> SchemaValidator:
+        """The pydantic-core SchemaValidator used to validate instances of the model."""
         try:
-            validator = _getattr_no_parents(type, '__pydantic_validator__')
+            existing_validator = _getattr_no_parents(self._type, '__pydantic_validator__')
+            if isinstance(existing_validator, SchemaValidator):
+                return existing_validator  # Do not return MockValSer as it causes another core schema to be built
```

Comment:
@sydney-runkle was this a "bug"? It used here possible `MockValSer` which caused double the core schema building. (Also happens in `master`). First core schema was constructed in `TypeAdapter.__init__` and second one in `MockValSer._attempt_rebuild()` when `TypeAdapter.validator` was used 🤔 So this happened when the model used `defer_build=True`

I now fixed it here as I noticed the strange duplicated core schema building in the test I added here https://github.com/pydantic/pydantic/pull/8939/files#diff-da24b355349b53e4c8047034767d008e4c2ce9ea6c8933a3acd50d0828f00528R766
That assertion failed with the duplication but its now fixed

Reply:
This probably explains why we saw even worse performance when trying out `defer_build` as it actually caused even more work to happen in the service. So the heavy core schemas being built 2x with `defer_build`.